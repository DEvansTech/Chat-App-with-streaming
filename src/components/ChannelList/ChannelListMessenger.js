import React, { PureComponent } from 'react';
import { FlatList } from 'react-native';
import PropTypes from 'prop-types';
import { ChannelPreview, ChannelPreviewMessenger } from '../ChannelPreview';

import { withChatContext } from '../../context';

import ChannelListHeaderNetworkDownIndicator from './ChannelListHeaderNetworkDownIndicator';
import ChannelListHeaderErrorIndicator from './ChannelListHeaderErrorIndicator';
import ChannelListFooterLoadingIndicator from './ChannelListFooterLoadingIndicator';

import {
  LoadingIndicator,
  LoadingErrorIndicator,
  EmptyStateIndicator,
} from '../Indicators';

/**
 * ChannelListMessenger - UI component for list of channels, allowing you to select the channel you want to open
 *
 * @example ../docs/ChannelListMessenger.md
 */
class ChannelListMessenger extends PureComponent {
  static propTypes = {
    /** Channels can be either an array of channels or a promise which resolves to an array of channels */
    channels: PropTypes.oneOfType([
      PropTypes.array,
      PropTypes.objectOf({
        then: PropTypes.func,
      }),
      PropTypes.object,
    ]).isRequired,
    setActiveChannel: PropTypes.func,
    /** UI Component to display individual channel item in list.
     * Defaults to [ChannelPreviewMessenger](https://getstream.github.io/stream-chat-react-native/#channelpreviewmessenger) */
    Preview: PropTypes.oneOfType([PropTypes.node, PropTypes.elementType]),
    /** The loading indicator to use. Default: [LoadingIndicator](https://getstream.github.io/stream-chat-react-native/#loadingindicator) */
    LoadingIndicator: PropTypes.oneOfType([
      PropTypes.node,
      PropTypes.elementType,
    ]),
    /** The indicator to use when there is error in fetching channels. Default: [LoadingErrorIndicator](https://getstream.github.io/stream-chat-react-native/#loadingerrorindicator) */
    LoadingErrorIndicator: PropTypes.oneOfType([
      PropTypes.node,
      PropTypes.elementType,
    ]),
    /** The indicator to use when channel list is empty. Default: [EmptyStateIndicator](https://getstream.github.io/stream-chat-react-native/#emptystateindicator) */
    EmptyStateIndicator: PropTypes.oneOfType([
      PropTypes.node,
      PropTypes.elementType,
    ]),
    /**
     * The indicator to display network-down error at top of list, if there is connectivity issue
     * Default: [ChannelListHeaderNetworkDownIndicator](https://getstream.github.io/stream-chat-react-native/#ChannelListHeaderNetworkDownIndicator)
     */
    HeaderNetworkDownIndicator: PropTypes.oneOfType([
      PropTypes.node,
      PropTypes.elementType,
    ]),
    /**
     * The indicator to display error at top of list, if there was an error loading some page/channels after the first page.
     * Default: [ChannelListHeaderErrorIndicator](https://getstream.github.io/stream-chat-react-native/#ChannelListHeaderErrorIndicator)
     */
    HeaderErrorIndicator: PropTypes.oneOfType([
      PropTypes.node,
      PropTypes.elementType,
    ]),
    /**
     * Loading indicator to display at bottom of the list, while loading further pages.
     * Default: [ChannelListFooterLoadingIndicator](https://getstream.github.io/stream-chat-react-native/#ChannelListFooterLoadingIndicator)
     */
    FooterLoadingIndicator: PropTypes.oneOfType([
      PropTypes.node,
      PropTypes.elementType,
    ]),
    /** Remove all the existing channels from UI and load fresh channels. */
    reloadList: PropTypes.func,
    /** Loads next page of channels in channels object, which is present here as prop */
    loadNextPage: PropTypes.func,
    /**
     * Refresh the channel list. Its similar to `reloadList`, but it doesn't wipe out existing channels
     * from UI before loading new set of channels.
     */
    refreshList: PropTypes.func,
    /**
     * For flatlist
     * @see See loeadMoreThreshold [doc](https://facebook.github.io/react-native/docs/flatlist#onendreachedthreshold)
     * */
    loadMoreThreshold: PropTypes.number,
    /** If there is error in querying channels */
    error: PropTypes.oneOfType([PropTypes.bool, PropTypes.object]),
    /** If channels are being queries. LoadingIndicator will be displayed if true */
    loadingChannels: PropTypes.bool,
    /** If channel list is being refreshed. Loader at top of the list will be displayed if true. */
    refreshing: PropTypes.bool,
    /** If further channels are being loadded. Loader will be shown at bottom of the list */
    loadingNextPage: PropTypes.bool,
    /**
     * Besides existing (default) UX behaviour of underlying flatlist of ChannelListMessenger component, if you want
     * to attach some additional props to un derlying flatlist, you can add it to following prop.
     *
     * You can find list of all the available FlatList props here - https://facebook.github.io/react-native/docs/flatlist#props
     *
     * **NOTE** Don't use `additionalFlatListProps` to get access to ref of flatlist. Use `setFlatListRef` instead.
     *
     * e.g.
     * ```
     * <ChannelListMessenger
     *  channels={channels}
     *  additionalFlatListProps={{ bounces: true }} />
     * ```
     */
    additionalFlatListProps: PropTypes.object,
    /**
     * Use `setFlatListRef` to get access to ref to inner FlatList.
     *
     * e.g.
     * ```
     * <ChannelListMessenger
     *  setFlatListRef={(ref) => {
     *    // Use ref for your own good
     *  }}
     * ```
     */
    setFlatListRef: PropTypes.func,
  };

  static defaultProps = {
    Preview: ChannelPreviewMessenger,
    LoadingIndicator,
    LoadingErrorIndicator,
    HeaderNetworkDownIndicator: ChannelListHeaderNetworkDownIndicator,
    HeaderErrorIndicator: ChannelListHeaderErrorIndicator,
    FooterLoadingIndicator: ChannelListFooterLoadingIndicator,
    EmptyStateIndicator,
    // https://github.com/facebook/react-native/blob/a7a7970e543959e9db5281914d5f132beb01db8d/Libraries/Lists/VirtualizedList.js#L466
    loadMoreThreshold: 2,
    additionalFlatListProps: {},
  };

  renderLoading = () => {
    const Indicator = this.props.LoadingIndicator;
    return <Indicator listType='channel' />;
  };

  renderLoadingError = () => {
    const Indicator = this.props.LoadingErrorIndicator;
    return (
      <Indicator
        error={this.props.error}
        retry={this.props.reloadList}
        listType='channel'
        loadNextPage={this.props.loadNextPage}
      />
    );
  };

  renderEmptyState = () => {
    const Indicator = this.props.EmptyStateIndicator;
    return <Indicator listType='channel' />;
  };

  renderHeaderIndicator = () => {
    const { isOnline, error, refreshList } = this.props;

    if (!isOnline) {
      const HeaderNetworkDownIndicator = this.props.HeaderNetworkDownIndicator;
      return <HeaderNetworkDownIndicator />;
    }

    if (error) {
      const HeaderErrorIndicator = this.props.HeaderErrorIndicator;
      return <HeaderErrorIndicator onPress={refreshList} />;
    }
  };

  renderChannels = () => (
    <>
      {this.renderHeaderIndicator()}
      <FlatList
        ref={(flRef) => {
          this.props.setFlatListRef && this.props.setFlatListRef(flRef);
        }}
        data={this.props.channels}
        onEndReached={() => this.props.loadNextPage(false)}
        onRefresh={() => this.props.refreshList()}
        refreshing={this.props.refreshing}
        onEndReachedThreshold={this.props.loadMoreThreshold}
        ListEmptyComponent={this.renderEmptyState}
        ListFooterComponent={() => {
          if (this.props.loadingNextPage) {
            const FooterLoadingIndicator = this.props.FooterLoadingIndicator;

            return <FooterLoadingIndicator />;
          }

          return null;
        }}
        renderItem={({ item: channel }) => (
          <ChannelPreview
            {...this.props}
            key={channel.cid}
            channel={channel}
            Preview={this.props.Preview}
          />
        )}
        keyExtractor={(item) => item.cid}
        {...this.props.additionalFlatListProps}
      />
    </>
  );

  render() {
    if (this.props.error && this.props.channels.length === 0) {
      return this.renderLoadingError();
    } else if (this.props.loadingChannels) {
      return this.renderLoading();
    } else {
      return this.renderChannels();
    }
  }
}

export default withChatContext(ChannelListMessenger);
