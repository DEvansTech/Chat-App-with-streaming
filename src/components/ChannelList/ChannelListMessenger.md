Unread channel preview:

```js
const data = require('../docs/data.js');
const View = require('react-native').View;

<View
  style={{
    height: '500px',
  }}
>
  <Chat client={data.client}>
    <ChannelList channels={data.channels} List={ChannelListMessenger} />
  </Chat>
</View>;
```
