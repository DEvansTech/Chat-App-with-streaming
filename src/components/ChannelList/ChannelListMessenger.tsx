import React from 'react';
import { FlatList } from 'react-native';
import type { Channel, UnknownType } from 'stream-chat';

import ChannelListFooterLoadingIndicator from './ChannelListFooterLoadingIndicator';
import ChannelListHeaderErrorIndicator from './ChannelListHeaderErrorIndicator';
import ChannelListHeaderNetworkDownIndicator from './ChannelListHeaderNetworkDownIndicator';
import ChannelPreview from '../ChannelPreview/ChannelPreview';
import EmptyStateIndicatorDefault from '../Indicators/EmptyStateIndicator';
import LoadingErrorIndicatorDefault from '../Indicators/LoadingErrorIndicator';
import LoadingIndicatorDefault from '../Indicators/LoadingIndicator';

import { useChatContext } from '../../contexts/chatContext/ChatContext';

import type { ChannelListProps } from './ChannelList';

import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
} from '../../types/types';

export type ChannelListMessengerProps<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
> = ChannelListProps<At, Ch, Co, Ev, Me, Re, Us> & {
  /**
   * Channels can be either an array of channels or a promise which resolves to an array of channels
   */
  channels: Channel<At, Ch, Co, Ev, Me, Re, Us>[];
  /**
   * Error in channels query, if any
   * */
  error?: boolean;
  /**
   * Incremental number change to force update the FlatList
   */
  forceUpdate?: number;
  /**
   * Whether or not the FlatList has another page to render
   */
  hasNextPage?: boolean;
  /**
   * Initial channels query loading state, triggers the LoadingIndicator
   */
  loadingChannels?: boolean;
  /**
   * Whether or not additional channels are being loaded, triggers the FooterLoadingIndicator
   * */
  loadingNextPage?: boolean;
  /**
   * Loads the next page of `channels`, which is present as a required prop
   * */
  loadNextPage?: () => Promise<void> | null;
  /**
   * Triggered when the channel list is refreshing, displays a loading spinner at the top of the list
   * */
  refreshing?: boolean;
  /**
   * Function to refresh the channel list that is similar to `reloadList`, but it doesn't wipe out existing channels
   * from UI before loading the new ones
   */
  refreshList?: () => Promise<void>;

  /**
   * Removes all the existing channels from UI and loads fresh channels
   * */
  reloadList?: () => Promise<void>;
  /**
   * Function to set the currently active channel, acts as a bridge between ChannelList and Channel components
   *
   * @param channel A channel object
   * */
  setActiveChannel?: (channel: Channel<At, Ch, Co, Ev, Me, Re, Us>) => void;
  /**
   * Function to gain access to the inner FlatList ref
   *
   * **Example:**
   *
   * ```
   * <ChannelListMessenger
   *  setFlatListRef={(ref) => {
   *    // Use ref for your own good
   *  }}
   * ```
   */
  setFlatListRef?: (
    ref: FlatList<Channel<At, Ch, Co, Ev, Me, Re, Us>> | null,
  ) => void;
};

/**
 * This UI component displays the preview list of channels and handles Channel navigation.
 *
 * @example ../docs/ChannelListMessenger.md
 */
const ChannelListMessenger = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  props: ChannelListMessengerProps<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const {
    additionalFlatListProps = {},
    channels,
    EmptyStateIndicator = EmptyStateIndicatorDefault,
    error,
    FooterLoadingIndicator = ChannelListFooterLoadingIndicator,
    forceUpdate,
    loadingChannels,
    LoadingErrorIndicator = LoadingErrorIndicatorDefault,
    LoadingIndicator = LoadingIndicatorDefault,
    loadingNextPage,
    // https://github.com/facebook/react-native/blob/a7a7970e543959e9db5281914d5f132beb01db8d/Libraries/Lists/VirtualizedList.js#L466
    loadMoreThreshold = 2,
    loadNextPage,
    refreshing,
    refreshList,
    reloadList,
    setFlatListRef,
  } = props;

  const { isOnline } = useChatContext<At, Ch, Co, Ev, Me, Re, Us>();

  const renderHeaderIndicator = () => {
    const {
      HeaderErrorIndicator = ChannelListHeaderErrorIndicator,
      HeaderNetworkDownIndicator = ChannelListHeaderNetworkDownIndicator,
    } = props;
    if (!isOnline) {
      return <HeaderNetworkDownIndicator />;
    } else if (error) {
      return <HeaderErrorIndicator onPress={refreshList} />;
    }
    return null;
  };

  const renderItem = (channel: Channel<At, Ch, Co, Ev, Me, Re, Us>) => (
    <ChannelPreview<At, Ch, Co, Ev, Me, Re, Us> {...props} channel={channel} />
  );

  const renderChannels = () => (
    <>
      {renderHeaderIndicator()}
      <FlatList
        data={channels}
        extraData={forceUpdate}
        keyExtractor={(item) => item.cid}
        ListEmptyComponent={() => <EmptyStateIndicator listType='channel' />}
        ListFooterComponent={() =>
          loadingNextPage ? <FooterLoadingIndicator /> : null
        }
        onEndReached={loadNextPage}
        onEndReachedThreshold={loadMoreThreshold}
        onRefresh={refreshList}
        ref={(flRef) => {
          setFlatListRef && setFlatListRef(flRef);
        }}
        refreshing={refreshing}
        renderItem={({ item }) => renderItem(item)}
        testID='channel-list-messenger'
        {...additionalFlatListProps}
      />
    </>
  );

  if (error && !refreshing && !(channels && channels.length)) {
    return (
      <LoadingErrorIndicator
        error={error}
        listType='channel'
        loadNextPage={loadNextPage}
        retry={reloadList}
      />
    );
  } else if (loadingChannels) {
    return <LoadingIndicator listType='channel' />;
  } else {
    return renderChannels();
  }
};

export default ChannelListMessenger;
