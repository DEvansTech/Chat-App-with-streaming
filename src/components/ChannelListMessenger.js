import React, { PureComponent } from 'react';
import { FlatList } from 'react-native';
import PropTypes from 'prop-types';
import { ChannelPreview } from './ChannelPreview';
import { ChannelPreviewMessenger } from './ChannelPreviewMessenger';
import { withChatContext } from '../context';

import { LoadingIndicator } from './LoadingIndicator';
import { LoadingErrorIndicator } from './LoadingErrorIndicator';
import { EmptyStateIndicator } from './EmptyStateIndicator';

/**
 * ChannelListMessenger - UI component for list of channels, allowing you to select the channel you want to open
 *
 * @example ./docs/ChannelListMessenger.md
 */
const ChannelListMessenger = withChatContext(
  class ChannelListMessenger extends PureComponent {
    static propTypes = {
      /** Channels can be either an array of channels or a promise which resolves to an array of channels */
      channels: PropTypes.oneOfType([
        PropTypes.array,
        PropTypes.objectOf({
          then: PropTypes.func,
        }),
        PropTypes.object,
      ]).isRequired,
      setActiveChannel: PropTypes.func,
      /** UI Component to display individual channel item in list.
       * Defaults to [ChannelPreviewMessenger](https://getstream.github.io/stream-chat-react-native/#channelpreviewmessenger) */
      Preview: PropTypes.oneOfType([PropTypes.node, PropTypes.elementType]),
      /** The loading indicator to use. Default: [LoadingIndicator](https://getstream.github.io/stream-chat-react-native/#loadingindicator) */
      LoadingIndicator: PropTypes.oneOfType([
        PropTypes.node,
        PropTypes.elementType,
      ]),
      /** The indicator to use when there is error in fetching channels. Default: [LoadingErrorIndicator](https://getstream.github.io/stream-chat-react-native/#loadingerrorindicator) */
      LoadingErrorIndicator: PropTypes.oneOfType([
        PropTypes.node,
        PropTypes.elementType,
      ]),
      /** The indicator to use when channel list is empty. Default: [EmptyStateIndicator](https://getstream.github.io/stream-chat-react-native/#emptystateindicator) */
      EmptyStateIndicator: PropTypes.oneOfType([
        PropTypes.node,
        PropTypes.elementType,
      ]),
      /** Loads next page of channels in channels object, which is present here as prop */
      loadNextPage: PropTypes.func,
      /**
       * For flatlist
       * @see See loeadMoreThreshold [doc](https://facebook.github.io/react-native/docs/flatlist#onendreachedthreshold)
       * */
      loadMoreThreshold: PropTypes.number,
      /** If there is error in querying channels */
      error: PropTypes.oneOfType([PropTypes.bool, PropTypes.object]),
      /** If channels are being queries. LoadingIndicator will be displayed if true */
      loadingChannels: PropTypes.bool,
      /**
       * Besides existing (default) UX behaviour of underlying flatlist of ChannelListMessenger component, if you want
       * to attach some additional props to un derlying flatlist, you can add it to following prop.
       *
       * You can find list of all the available FlatList props here - https://facebook.github.io/react-native/docs/flatlist#props
       *
       * **NOTE** Don't use `additionalFlatListProps` to get access to ref of flatlist. Use `setFlatListRef` instead.
       *
       * e.g.
       * ```
       * <ChannelListMessenger
       *  channels={channels}
       *  additionalFlatListProps={{ bounces: true }} />
       * ```
       */
      additionalFlatListProps: PropTypes.object,
      /**
       * Use `setFlatListRef` to get access to ref to inner FlatList.
       *
       * e.g.
       * ```
       * <ChannelListMessenger
       *  setFlatListRef={(ref) => {
       *    // Use ref for your own good
       *  }}
       * ```
       */
      setFlatListRef: PropTypes.func,
    };

    static defaultProps = {
      Preview: ChannelPreviewMessenger,
      LoadingIndicator,
      LoadingErrorIndicator,
      EmptyStateIndicator,
      // https://github.com/facebook/react-native/blob/a7a7970e543959e9db5281914d5f132beb01db8d/Libraries/Lists/VirtualizedList.js#L466
      loadMoreThreshold: 2,
      additionalFlatListProps: {},
    };

    renderLoading = () => {
      const Indicator = this.props.LoadingIndicator;
      return <Indicator listType="channel" />;
    };

    renderLoadingError = () => {
      const Indicator = this.props.LoadingErrorIndicator;
      return <Indicator error={this.props.error} listType="channel" />;
    };

    renderEmptyState = () => {
      const Indicator = this.props.EmptyStateIndicator;
      return <Indicator listType="channel" />;
    };

    renderChannels = () => (
      <FlatList
        ref={(flRef) => {
          this.props.setFlatListRef && this.props.setFlatListRef(flRef);
        }}
        data={this.props.channels}
        onEndReached={this.props.loadNextPage}
        onEndReachedThreshold={this.props.loadMoreThreshold}
        ListEmptyComponent={this.renderEmptyState}
        renderItem={({ item: channel }) => (
          <ChannelPreview
            {...this.props}
            key={channel.cid}
            channel={channel}
            Preview={this.props.Preview}
          />
        )}
        keyExtractor={(item) => item.cid}
        {...this.props.additionalFlatListProps}
      />
    );

    render() {
      if (this.props.error) {
        return this.renderLoadingError();
      } else if (this.props.loadingChannels) {
        return this.renderLoading();
      } else {
        return this.renderChannels();
      }
    }
  },
);

export { ChannelListMessenger };
