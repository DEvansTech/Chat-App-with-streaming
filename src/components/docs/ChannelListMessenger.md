Unread channel preview:

```js
const data = require('./data');
const View = require('react-native').View;

<View
  style={{
    height: '500px',
  }}
>
  <Chat client={data.client}>
    <ChannelList channels={data.channels} List={ChannelListMessenger} />
  </Chat>
</View>;
```
