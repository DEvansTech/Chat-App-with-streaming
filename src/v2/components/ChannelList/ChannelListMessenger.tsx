import React, { useEffect, useState } from 'react';
import { RefreshControl, StyleSheet, View } from 'react-native';
import { FlatList } from 'react-native-gesture-handler';

import { ChannelPreview } from '../ChannelPreview/ChannelPreview';

import { useChatContext } from '../../contexts/chatContext/ChatContext';
import {
  ChannelsContextValue,
  useChannelsContext,
} from '../../contexts/channelsContext/ChannelsContext';
import { useTheme } from '../../contexts/themeContext/ThemeContext';

import type { Channel } from 'stream-chat';

import type {
  DefaultAttachmentType,
  DefaultChannelType,
  DefaultCommandType,
  DefaultEventType,
  DefaultMessageType,
  DefaultReactionType,
  DefaultUserType,
  UnknownType,
} from '../../types/types';

const styles = StyleSheet.create({
  flatList: { flexGrow: 1, flexShrink: 1 },
  flatListContentContainer: { flexGrow: 1 },
  header: {
    alignItems: 'center',
    flex: 1,
    justifyContent: 'center',
    padding: 8,
  },
  searchContainer: {
    alignItems: 'center',
    borderRadius: 18,
    borderWidth: 1,
    flexDirection: 'row',
    paddingHorizontal: 10,
    paddingVertical: 8,
    width: '100%',
  },
  searchInput: { paddingLeft: 16 },
  statusIndicator: { left: 0, position: 'absolute', right: 0, top: 0 },
});

export type ChannelListMessengerPropsWithContext<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
> = Omit<
  ChannelsContextValue<At, Ch, Co, Ev, Me, Re, Us>,
  | 'hasNextPage'
  | 'HeaderErrorIndicator'
  | 'HeaderNetworkDownIndicator'
  | 'onSelect'
  | 'Preview'
>;

const StatusIndicator = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>() => {
  const { isOnline } = useChatContext<At, Ch, Co, Ev, Me, Re, Us>();
  const {
    error,
    HeaderErrorIndicator,
    HeaderNetworkDownIndicator,
    loadingChannels,
    refreshList,
  } = useChannelsContext<At, Ch, Co, Ev, Me, Re, Us>();

  if (loadingChannels) return null;

  if (!isOnline) {
    return (
      <View style={styles.statusIndicator}>
        <HeaderNetworkDownIndicator />
      </View>
    );
  } else if (error) {
    return (
      <View style={styles.statusIndicator}>
        <HeaderErrorIndicator onPress={refreshList} />
      </View>
    );
  }
  return null;
};

const ChannelListMessengerWithContext = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  props: ChannelListMessengerPropsWithContext<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const {
    additionalFlatListProps,
    channels,
    EmptyStateIndicator,
    error,
    FooterLoadingIndicator,
    forceUpdate,
    ListHeaderComponent,
    loadingChannels,
    LoadingErrorIndicator,
    loadingNextPage,
    loadMoreThreshold,
    loadNextPage,
    numberOfSkeletons,
    refreshing,
    refreshList,
    reloadList,
    setFlatListRef,
    Skeleton,
  } = props;

  const {
    theme: {
      channelListMessenger: { flatList, flatListContent },
    },
  } = useTheme();

  /**
   * In order to prevent the EmptyStateIndicator component from showing up briefly on mount,
   * we set the loading state one cycle behind to ensure the channels are set before the
   * change to loadingChannels is registered.
   */
  const [loading, setLoading] = useState(true);
  useEffect(() => {
    if (!!loadingChannels !== loading) {
      setLoading(!!loadingChannels);
    }
  }, [loading, loadingChannels]);

  const renderItem = (channel: Channel<At, Ch, Co, Ev, Me, Re, Us>) => (
    <ChannelPreview<At, Ch, Co, Ev, Me, Re, Us> channel={channel} />
  );

  if (error && !refreshing && !loadingChannels && !channels?.length) {
    return (
      <LoadingErrorIndicator
        error={error}
        listType='channel'
        loadNextPage={loadNextPage}
        retry={reloadList}
      />
    );
  }

  return (
    <>
      <FlatList
        contentContainerStyle={[
          styles.flatListContentContainer,
          flatListContent,
        ]}
        data={channels}
        extraData={forceUpdate}
        keyExtractor={(item) => item.cid}
        ListEmptyComponent={
          loading ? (
            <View>
              {Array.from(Array(numberOfSkeletons)).map((_, index) => (
                <Skeleton key={`skeleton_${index}`} />
              ))}
            </View>
          ) : (
            <EmptyStateIndicator listType='channel' />
          )
        }
        ListFooterComponent={
          loadingNextPage ? <FooterLoadingIndicator /> : undefined
        }
        ListHeaderComponent={() =>
          channels.length && ListHeaderComponent ? (
            <ListHeaderComponent />
          ) : null
        }
        onEndReached={() => {
          if (loadNextPage) {
            loadNextPage();
          }
        }}
        onEndReachedThreshold={loadMoreThreshold}
        ref={setFlatListRef}
        refreshControl={
          <RefreshControl onRefresh={refreshList} refreshing={refreshing} />
        }
        renderItem={({ item }) => renderItem(item)}
        style={[styles.flatList, flatList]}
        testID='channel-list-messenger'
        {...additionalFlatListProps}
      />
      <StatusIndicator<At, Ch, Co, Ev, Me, Re, Us> />
    </>
  );
};

export type ChannelListMessengerProps<
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
> = Partial<ChannelListMessengerPropsWithContext<At, Ch, Co, Ev, Me, Re, Us>>;

/**
 * This UI component displays the preview list of channels and handles Channel navigation. It
 * receives all props from the ChannelList component.
 *
 * @example ./ChannelListMessenger.md
 */
export const ChannelListMessenger = <
  At extends UnknownType = DefaultAttachmentType,
  Ch extends UnknownType = DefaultChannelType,
  Co extends string = DefaultCommandType,
  Ev extends UnknownType = DefaultEventType,
  Me extends UnknownType = DefaultMessageType,
  Re extends UnknownType = DefaultReactionType,
  Us extends UnknownType = DefaultUserType
>(
  props: ChannelListMessengerProps<At, Ch, Co, Ev, Me, Re, Us>,
) => {
  const {
    additionalFlatListProps,
    channels,
    EmptyStateIndicator,
    error,
    FooterLoadingIndicator,
    forceUpdate,
    ListHeaderComponent,
    loadingChannels,
    LoadingErrorIndicator,
    LoadingIndicator,
    loadingNextPage,
    loadMoreThreshold,
    loadNextPage,
    numberOfSkeletons,
    refreshing,
    refreshList,
    reloadList,
    setFlatListRef,
    Skeleton,
  } = useChannelsContext<At, Ch, Co, Ev, Me, Re, Us>();

  return (
    <ChannelListMessengerWithContext
      {...{
        additionalFlatListProps,
        channels,
        EmptyStateIndicator,
        error,
        FooterLoadingIndicator,
        forceUpdate,
        ListHeaderComponent,
        loadingChannels,
        LoadingErrorIndicator,
        LoadingIndicator,
        loadingNextPage,
        loadMoreThreshold,
        loadNextPage,
        numberOfSkeletons,
        refreshing,
        refreshList,
        reloadList,
        setFlatListRef,
        Skeleton,
      }}
      {...props}
    />
  );
};

ChannelListMessenger.displayName = 'ChannelListMessenger{channelListMessenger}';
