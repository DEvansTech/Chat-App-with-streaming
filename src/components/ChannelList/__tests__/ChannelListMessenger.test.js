import React from 'react';
import { Text, View } from 'react-native';
import { cleanup } from '@testing-library/react-native';
import renderer from 'react-test-renderer';

import ChannelListMessenger from '../ChannelListMessenger';

const channels = [
  { id: '1', name: 'dan' },
  { id: '2', name: 'neil' },
];

const ComponentTest = () => (
  <>
    <View>
      <Text>{channels[0].name}</Text>
    </View>
    <View>
      <Text>{channels[1].name}</Text>
    </View>
  </>
);

const Component = ({
  error = false,
  loadingChannels = false,
  refreshing = false,
}) => (
  <ChannelListMessenger
    channels={channels}
    error={error}
    loadingChannels={loadingChannels}
    LoadingErrorIndicator={() => (
      <View>
        <Text>Loading Error Indicator</Text>
      </View>
    )}
    LoadingIndicator={() => (
      <View>
        <Text>Loading Indicator</Text>
      </View>
    )}
    refreshing={refreshing}
  />
);

describe('ChannelListMessenger', () => {
  afterEach(cleanup);

  it('passes', () => {
    expect(true).toBeTruthy();
  });

  it('renders a test component correctly', () => {
    const tree = renderer.create(<ComponentTest />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders correctly', () => {
    const tree = renderer.create(<Component />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  // it('when `error` prop is true, `LoadingErrorIndicator` should be rendered', () => {
  //   const tree = renderer.create(<Component error={true} />).toJSON();
  //   expect(tree).toMatchSnapshot();
  // });

  // it('when `loading` prop is true, `LoadingIndicator` should be rendered', () => {
  //   const tree = renderer.create(<Component loading={true} />).toJSON();
  //   expect(tree).toMatchSnapshot();
  // });
});
