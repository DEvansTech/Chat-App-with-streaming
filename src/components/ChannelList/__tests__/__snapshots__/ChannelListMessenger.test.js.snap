// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChannelListMessenger renders a test component correctly 1`] = `
Array [
  <View>
    <Text>
      dan
    </Text>
  </View>,
  <View>
    <Text>
      neil
    </Text>
  </View>,
]
`;
